### PR TITLE
New profiler

### PR DESCRIPTION
**Pull request by Mr-Wiseguy**
Replaces the puppyprint profiler and can be used independently of puppyprint debug if desired. More accurate, more responsive, more detailed, and can be tuned to average samples over large periods of time without any performance impact.

### CODE REVIEW

**Comment by Mr-Wiseguy** [comment]
Standalone mode: image


Integrated into puppyprint debug: 
image

File: src/game/area.c [hunk 1]
```diff
@@ -434,11 +431,11 @@ void render_game(void) {
 
     gViewportOverride = NULL;
     gViewportClip     = NULL;
-
+    
+    fast_profiler_update(PROFILER_TIME_GFX);
+    // TODO wiseguy move this to puppyprint if puppyprint is enabled
+    fast_profiler_print_times();
```
Comment (thecozies) [inline src/game/area.c:437]: when you gonna do this todo

File: src/game/fasttext.c [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+  } else if (letter > '?') {
+    idx -= (1 + 3);
+  } else if (letter > ';') {
+    idx -= (3);
+  }
+
+  return (idx - TEX_ASCII_START) * 8;
+}
+
+static const u8 fasttest_font_kerning[] = {
```
Comment (thecozies) [inline src/game/fasttext.c:45]: fast test or fastest?

File: src/game/profiling.c [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+}
+
+void fast_profiler_rsp_resumed() {
+    rsp_pending_times[PROFILER_RSP_GFX] = osGetCount() - rsp_pending_times[PROFILER_RSP_GFX];
+}
+
+// This ends up being the same math as resumed, so we just use resumed for both
+// void fast_profiler_rsp_yielded() {
+//     rsp_pending_times[PROFILER_RSP_GFX] = osGetCount() - rsp_pending_times[PROFILER_RSP_GFX];
+// }
```
Comment (thecozies) [inline src/game/profiling.c:82]: whys this commented out?

File: src/game/object_list_processor.c [hunk 4]
```diff
@@ -559,7 +563,13 @@ void update_non_terrain_objects(void) {
 
     s32 i = 2;
     while ((listIndex = sObjectListUpdateOrder[i]) != -1) {
+        if (listIndex == OBJ_LIST_PLAYER) {
+            fast_profiler_update(PROFILER_TIME_BEHAVIOR1);
+        }
         gObjectCounter += update_objects_in_list(&gObjectLists[listIndex]);
+        if (listIndex == OBJ_LIST_PLAYER) {
+            fast_profiler_update(PROFILER_TIME_MARIO);
+        }
```
Comment (thecozies) [inline src/game/object_list_processor.c:572]: are these if statements compiled out if fast_profiler_update is a define noop?

**Review by thecozies — COMMENTED** [review]
few comments and thoughts (i still need to do testing)

- looks like we're losing specific CPU tests for collision, and possibly graph node specific tests? we could always add important ones back but thought it was worth mentioning. i think its good to keep an eye on those areas but maybe theres an easier way we can wrap specific functions for profiling and we can improve it without the bloat.
- i know your profiler is fast but can you find/replace `fast_profiler` with `profiler` lmao it seems excessive, especially coming from profiling.c and not fast_profiling.c
- behavior1 vs. behavior2, whats the dealio happening here? they seem kinda random at first glance

File: src/game/area.c (hunk 1 above)
Comment (Mr-Wiseguy) [reply src/game/area.c:437]: It is done oops

File: src/game/fasttext.c (hunk 2 above)
Comment (Mr-Wiseguy) [reply src/game/fasttext.c:45]: Neither, supposed to be fasttext

File: src/game/profiling.c (hunk 3 above)
Comment (Mr-Wiseguy) [reply src/game/profiling.c:82]: The comment above it explains why

File: src/game/object_list_processor.c (hunk 4 above)
Comment (Mr-Wiseguy) [reply src/game/object_list_processor.c:572]: Yeah, check the header. If the define is disabled, the functions become macros that do nothing, so this compiles into nothing as well.

**Comment by Mr-Wiseguy** [comment]
* I can add some sort of accumulator for specific function times, the hard part is making sure that audio thread preemptions don't leak into the measured durations. I can probably reuse the same preemption time mechanism for it that I use for general profiling, but maybe to not delay 2.0.2 any longer it should be slated as a 2.1.0 feature?
* Done, the name came from when it was a standalone thing
* I've renamed them to BEHAVIOR_BEFORE_MARIO and BEHAVIOR_AFTER_MARIO to make it more clear what they're measuring under the hood. It's basically a way of just isolating Mario's behavior time specifically.

**Review by thecozies — APPROVED** [review]
lgtm. one thing formatting wise would be that `fasttext.c` is the only file that uses 2 spaces for indention. makes my eye twitch but isnt a blocker

and yeah for 2.1.0 it'd be very useful having an easy to use misc profiler!

**Pull request merged**